api/service/legacysync/syncing.go

```diff
@@ -1209,3 +1133,119 @@ func GetSyncingPort(nodePort string) string {
 	}
 	return ""
 }
+
+// syncStatusExpiration is the expiration time out of a sync status.
+// If last sync result in memory is before the expiration, the sync status
+// will be updated.
+const syncStatusExpiration = 6 * time.Second
+
+type (
+	syncStatus struct {
+		lastResult     SyncCheckResult
+		lastUpdateTime time.Time
+		lock           sync.RWMutex
+	}
+
+	SyncCheckResult struct {
+		IsInSync    bool
+		OtherHeight uint64
+		HeightDiff  uint64
+	}
+)
+
+func newSyncStatus() syncStatus {
+	return syncStatus{}
+}
+
+func (status *syncStatus) isStatusExpired() bool {
+	status.lock.RLock()
+	defer status.lock.RUnlock()
+
+	return time.Since(status.lastUpdateTime) > syncStatusExpiration
+}
+
+func (status *syncStatus) update(result SyncCheckResult) {
+	status.lock.Lock()
+	defer status.lock.Unlock()
+
+	status.lastUpdateTime = time.Now()
+	status.lastResult = result
+}
+
+func (status *syncStatus) get() SyncCheckResult {
+	status.lock.RLock()
+	defer status.lock.RUnlock()
+
+	return status.lastResult
+}
+
+// GetSyncStatus get the last sync status for other modules (E.g. RPC, explorer).
+// If the last sync result is not expired, return the sync result immediately.
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *StateSync) GetSyncStatus() SyncCheckResult {
+	if !ss.syncStatus.isStatusExpired() {
+		return ss.syncStatus.get()
+	}
+	// If the result is expired, query the latest sync result
+	result := ss.isInSync(false)
+	ss.syncStatus.update(result)
+	return result
+}
+
+// GetSyncStatusDoubleChecked return the sync status when enforcing a immediate query on DNS nodes
+// with a double check to avoid false alarm.
+func (ss *StateSync) GetSyncStatusDoubleChecked() SyncCheckResult {
+	result := ss.isInSync(true)
+	ss.syncStatus.update(result)
+	return result
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *StateSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := ss.getMaxPeerHeight(false)
+	lastHeight := ss.blockChain.CurrentBlock().NumberU64()
+	wasOutOfSync := lastHeight+inSyncThreshold < otherHeight1
+
+	if !doubleCheck {
+		heightDiff := otherHeight1 - lastHeight
+		if otherHeight1 < lastHeight {
+			heightDiff = 0 //
+		}
+		utils.Logger().Info().
+			Uint64("OtherHeight", otherHeight1).
+			Uint64("lastHeight", lastHeight).
+			Msg("[SYNC] Checking sync status")
+		return SyncCheckResult{
+			IsInSync:    !wasOutOfSync,
+			OtherHeight: otherHeight1,
+			HeightDiff:  heightDiff,
+		}
+	}
+	// double check the sync status after 1 second to confirm (avoid false alarm)
+	time.Sleep(1 * time.Second)
```

Comment:
will this sleep block the RPC response time?

Reply:
It will not block RPC. The multithreading is only happening when the cache is updated.